pkg/monitor/clientview/gatherer.go

```diff
@@ -0,0 +1,174 @@
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	routeclient "github.com/openshift/client-go/route/clientset/versioned"
+	"github.com/openshift/library-go/test/library/metrics"
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+)
+
+var scrapePeriod = 30 * time.Second
+
+func FetchEventIntervalsForRestClientError(ctx context.Context, config *rest.Config, startTime time.Time) ([]monitorapi.Interval, error) {
+	kubeClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	routeClient, err := routeclient.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	_, err = kubeClient.CoreV1().Namespaces().Get(ctx, "openshift-monitoring", metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return []monitorapi.Interval{}, nil
+	}
+
+	kubeSvc, err := kubeClient.CoreV1().Services(metav1.NamespaceDefault).Get(ctx, "kubernetes", metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve cluster IP of kubernetes.default.svc - %v", err)
+	}
+
+	client, err := metrics.NewPrometheusClient(ctx, kubeClient, routeClient)
+	if err != nil {
+		return nil, err
+	}
+
+	gatherer := &gatherer{start: startTime, client: client, serviceNetworkIP: kubeSvc.Spec.ClusterIP}
+
+	// TODO: build more robust range query - run multiple range
+	//  queries with smaller time range ex: {from - from+30m}
+	result, err := gatherer.query(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("metric query for rest client failed: %v", err)
+	}
+
+	return gatherer.parse(result)
+}
+
+type gatherer struct {
+	start            time.Time
+	client           prometheusv1.API
+	serviceNetworkIP string
+}
+
+func (g *gatherer) query(ctx context.Context) (prometheustypes.Value, error) {
+	// TODO: should we include 5xx? "429|5..|<error>"
+	query := `rest_client_requests_total{code =~ "429|5..|<error>"}`
+	result, warnings, err := g.client.QueryRange(ctx, query, prometheusv1.Range{
+		Start: g.start,
+		End:   time.Now(),
+		// data is scraped every 30s in OpenShift
+		Step: scrapePeriod,
+	})
+	if err != nil {
+		framework.Logf("[RestClientParser]: prometheus client returned error: %v", err)
+		return nil, err
+	}
+	if len(warnings) > 0 {
+		framework.Logf("[RestClientParser]: #### warnings \\n\\t%v\\n\", strings.Join(warningsForQuery, \"\\n\\t\"", warnings)
+	}
+
+	return result, nil
+}
+
+func (g *gatherer) parse(result prometheustypes.Value) ([]monitorapi.Interval, error) {
+	intervals := []monitorapi.Interval{}
+	if result.Type() != prometheustypes.ValMatrix {
+		return nil, fmt.Errorf("expected a Matrix")
+	}
+
+	matrix := result.(prometheustypes.Matrix)
+	for _, series := range matrix {
+		if len(series.Values) <= 1 {
+			continue
+		}
+		previous := series.Values[0].Value
+		for _, current := range series.Values[1:] {
+			if !previous.Equal(current.Value) {
```

Comment:
Am I correct that if we're experiencing client errors over a prolonged period of time, every metric sample that we increment results in it's own interval? How far apart are the timestamps on each value? Every 15s? 

I'm concerned about bounding here, if a bad job is going to generate a million intervals? That would break a bunch of things and cost us money likely too.

For example with disruption we record a single interval for a prolonged period of disruption, we record when we start seeing disruption (and what error message), and then watch for it to stop or for the error message to change. Either event results in the interval now having a To time, so we create it.

IMO you should apply similar here such that a prolonged uninterrupted period of is just one interval. Granted this may get complicated, maybe we have one value where it doesn't change and then it starts going up again, that would make a new interval, but that would be better than nothing. Depends on how far apart the prometheus samples are as well.

Reply:
> every metric sample that we increment results in it's own interval? 

No, not exactly. We're going to create one interval for every scrape period (30s) if number of client errors has increased.

> I'm concerned about bounding here, if a bad job is going to generate a million intervals?

No, one interval every 30s at most.

Reply:
Ok one ever 30s, but is that also per time series returned by your promql? (it looks like that might catch a fair bit) Example if that promql returns 500 time series and we have a bad job run with a 30min problem, thats 500 * 30 * 2 or 30k intervals, which would double the number I'd expect normally. 

If the error counter for a time series increments at t, t+30s, t+60s, t+90s, I think it's probably worth the effort to make that one interval as we do for disruption. That would involve tracking start points, watching for a sample that didn't increment and using that as a trigger to terminate that interval and add it to the list, and terminating any intervals that were dangling at the end of the job run.

Reply:
Its per time series, yes, but I don't expect it to be significant. Its "error code" * "source namespace", so in the worst case its 10 error codes * 50 namespaces * (30 mins job / 30 secs interval) = 250 intervals for "every component firing 10 different errors throught the whole test duration)

Reply:
Unless I'm missing something, 10 error codes x 50 namespaces x 3 api points x 60 (30 minutes at 30s intervals) = 90,000, not 750? I know it's an extreme example by even 10% of that is a lot more for our JS UI to handle. I think you need to batch based on consecutive failed samples.

Reply:
Ah, you're correct. Right, this needs interval batching

Reply:
Implemented, now sequential intervals are merged into one:
```
  Aug 23 08:39:20.111: INFO: [client-rest-error-serializer] adding new interval Aug 23 08:36:58.784 - 1s    E client/APIError source/service-network node/10.128.0.197 namespace/openshift-apiserver component/openshift-apiserver client observed an API error - previous=1 current=5 rest_client_requests_total{apiserver="openshift-apiserver", code="503", container="openshift-apiserver", endpoint="https", host="172.30.0.1:443", instance="10.128.0.197:8443", job="api", method="GET", namespace="openshift-apiserver", pod="apiserver-5785c87bf8-9skb4", prometheus="openshift-monitoring/k8s", service="api"}
...
  Aug 23 08:39:20.111: INFO: [client-rest-error-serializer] updated existing interval Aug 23 08:36:58.784 - 30s   E client/APIError source/service-network node/10.128.0.197 namespace/openshift-apiserver component/openshift-apiserver client observed an API error - previous=1 current=5 rest_client_requests_total{apiserver="openshift-apiserver", code="503", container="openshift-apiserver", endpoint="https", host="172.30.0.1:443", instance="10.128.0.197:8443", job="api", method="GET", namespace="openshift-apiserver", pod="apiserver-5785c87bf8-9skb4", prometheus="openshift-monitoring/k8s", service="api"}
```